go/test/endtoend/vreplication/cluster.go

```diff
@@ -106,6 +109,128 @@ func setTempVtDataRoot() string {
 	return vtdataroot
 }
 
+// setVtMySQLRoot creates the root directory if it does not exist
+// and saves the directory in the VT_MYSQL_ROOT OS env var.
+// mysqlctl will then look for the mysql related binaries in the
+// ./bin, ./sbin, and ./libexec subdirectories of VT_MYSQL_ROOT.
+func setVtMySQLRoot(mysqlRoot string) error {
+	if _, err := os.Stat(mysqlRoot); os.IsNotExist(err) {
+		os.Mkdir(mysqlRoot, 0700)
+	}
+	err := os.Setenv("VT_MYSQL_ROOT", mysqlRoot)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("VT_MYSQL_ROOT is %s\n", mysqlRoot)
+	return nil
+}
+
+// setDBFlavor sets the MYSQL_FLAVOR OS env var.
+// You should call this after calling setVtMySQLRoot() to ensure that the
+// correct flavor is used by mysqlctl based on the current mysqld version
+// in the path. If you don't do this then mysqlctl will use the incorrect
+// config/mycnf/<flavor>.cnf file and mysqld may fail to start.
+func setDBFlavor() error {
+	versionStr, err := mysqlctl.GetVersionString()
+	if err != nil {
+		return err
+	}
+	f, v, err := mysqlctl.ParseVersionString(versionStr)
+	if err != nil {
+		return err
+	}
+	flavor := fmt.Sprintf("%s%d%d", f, v.Major, v.Minor)
+	err = os.Setenv("MYSQL_FLAVOR", string(flavor))
+	if err != nil {
+		return err
+	}
+	fmt.Printf("MYSQL_FLAVOR is %s\n", string(flavor))
+	return nil
+}
+
+func unsetVtMySQLRoot() {
+	_ = os.Unsetenv("VT_MYSQL_ROOT")
+}
+
+func unsetDBFlavor() {
+	_ = os.Unsetenv("MYSQL_FLAVOR")
+}
+
+// getDBTypeVersionInUse checks the major DB version of the mysqld binary
+// that mysqlctl would currently use, e.g. 5.7 or 8.0 (in semantic versioning
+// this would be major.minor but in MySQL it's effectively the major version).
+func getDBTypeVersionInUse() (string, error) {
+	var dbTypeMajorVersion string
+	versionStr, err := mysqlctl.GetVersionString()
+	if err != nil {
+		return dbTypeMajorVersion, err
+	}
+	flavor, version, err := mysqlctl.ParseVersionString(versionStr)
+	if err != nil {
+		return dbTypeMajorVersion, err
+	}
+	majorVersion := fmt.Sprintf("%d.%d", version.Major, version.Minor)
+	if flavor == mysqlctl.FlavorMySQL || flavor == mysqlctl.FlavorPercona {
+		dbTypeMajorVersion = fmt.Sprintf("mysql-%s", majorVersion)
+	} else {
+		dbTypeMajorVersion = fmt.Sprintf("%s-%s", strings.ToLower(string(flavor)), majorVersion)
+	}
+	return dbTypeMajorVersion, nil
+}
+
+// downloadDBTypeVersion downloads a recent major version release build for the specified
+// DB type.
+// If the file already exists, it will not download it again.
```

Comment:
The tarball will already exist if the test is running on a runner that it ran earlier on, right? I guess the probability over time will be high this will be true. But is there a way to bake these into the images itself? I am wondering if this will add flakiness because of unpredictable download times and/or availability issues.

Reply:
Yes, if a runner is re-used it would already be there.

I don't believe we have any control over the images used in GH actions (AFAIK, they are the same images used for CodeSpaces).

We already rely on being able to download MySQL specific things, we download and install the MySQL repo. Then install the mysql packages using the repo. The tar download is *only* done in this cross version test and IMO does not add significant risk beyond what we already have. We have plenty of other tests that are legit flaky today, but I have seen no issues with this test. We can always address any seen flakiness later though.